src/main/scala/com/wavesplatform/history/HistoryWriterImpl.scala

```diff
@@ -55,42 +53,41 @@ class HistoryWriterImpl private(file: Option[File], val synchronizationToken: Re
   override def discardBlock(): Seq[Transaction] = write { implicit lock =>
     val h = height()
     val transactions =
-      Block.parseBytes(blockBodyByHeight.mutate(_.remove(h))).fold(_ => Seq.empty[Transaction], _.transactionData.map(_.transaction))
-    scoreByHeight.mutate(_.remove(h))
-    val vOpt = Option(blockIdByHeight.mutate(_.remove(h)))
-    vOpt.map(v => heightByBlockId.mutate(_.remove(v)))
-    db.commit()
+      Block.parseBytes(blockBytes(h).getOrElse(Array[Byte]())).fold(_ => Seq.empty[Transaction], _.transactionData.map(_.transaction))
+
+    delete(blockBodyByHeightKey(h), None)
+    delete(scoreByHeightKey(h), None)
+    get(blockBodyByHeightKey(h)).foreach(b => ByteStrCodec.decode(b).toOption.foreach(r => delete(heightByBlockIdKey(r.value), None)))
```

Comment:
blockBodyByHeightKey -> blockIdByHeightKey

Reply:
@augmentedeconomy this will help to fix the rollback issue.